src/features/calendar/components/index.tsx

```diff
@@ -13,12 +14,62 @@ export enum TimeScale {
   MONTH = 'month',
 }
 
+function isValidDate(date: Date) {
+  return date instanceof Date && !isNaN(date as unknown as number);
+}
+
+function getTimeScale(timeScaleStr: string) {
+  let timeScale = TimeScale.MONTH;
+  if (
+    timeScaleStr !== undefined &&
+    Object.values(TimeScale).includes(timeScaleStr as TimeScale)
+  ) {
+    timeScale = timeScaleStr as TimeScale;
+  }
+  return timeScale;
+}
+
+function getDateFromString(focusDateStr: string) {
+  let date = new Date();
+  if (focusDateStr) {
+    const d = new Date(focusDateStr);
+    if (isValidDate(d)) {
+      date = d;
+    }
+  }
+  return date;
+}
+
 const Calendar = () => {
-  const [focusDate, setFocusDate] = useState<Date>(new Date());
+  const router = useRouter();
+
+  const focusDateStr = router.query.focusDate as string;
+  const [focusDate, setFocusDate] = useState(getDateFromString(focusDateStr));
+
+  const timeScaleStr = router.query.timeScale as string;
   const [selectedTimeScale, setSelectedTimeScale] = useState<TimeScale>(
-    TimeScale.MONTH
+    getTimeScale(timeScaleStr)
   );
 
+  useEffect(() => {
+    setFocusDate(getDateFromString(focusDateStr));
+  }, [focusDateStr]);
+
+  useEffect(() => {
+    setSelectedTimeScale(getTimeScale(timeScaleStr));
+  }, [timeScaleStr]);
+
+  useEffect(() => {
+    router.query.focusDate = dayjs(focusDate).format('YYYY-MM-DD');
+    router.query.timeScale = selectedTimeScale;
+    router.push(router, undefined, { shallow: true });
+  }, [focusDate, selectedTimeScale]);
+
+  function navigateTo(timeScale: TimeScale, date: Date) {
+    setSelectedTimeScale(timeScale);
+    setFocusDate(date);
+  }
```

Comment:
Is it possible to use the router as the data source, instead of keeping state in this component and trying to sync it with the router?

Reply:
Definitely possible, but when I tried it, it felt very slow on my laptop: it took about one second to switch between months or timescales. I didn't try it outside of dev mode, though.

Reply:
Hmm, that's interesting! I would not have thought that would affect performance!